### PR TITLE
EnsuresPredicateFix

### PR DESCRIPTION
**Pull request by AndreSonntag**
# Description

This pull request introduces the quickfix ensuresPredicate. The quickfix inserts the call "ensuresPredicate(predicate, variable)" above the error marker line. Currently, the quickfix is disabled, to activate it the comment out code in the quickfixer class must be enabled. 

The quickfix adds the CogniCrypt.jar to the client project. This helper jar is necessary to insert the ensuresPredicate() into the client code.

Structure CogniCrypt.jar:
```
package de.cognicrypt.staticanalyzer;
public class CC {
 public static boolean ensuresPredicate(String var,String predicate) {
	 return true;
 }
}
```
Images:
image

image

Fixes # (issue)
CogniCrypt side for Handling of reversal operations in CrySL and CogniCrypt(_SAST) #[80](https://github.com/CROSSINGTUD/CryptoAnalysis/issues/80)

## Type of change
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/EnsuresPredicateFix.java [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+		public ErrorSourceVisitor(final int lineNumber, final CompilationUnit unit, final ICompilationUnit sourceUnit) {
+			this.lineNumber = lineNumber;
+			this.unit = unit;
+			this.sourceUnit = sourceUnit;
+		}
+
+		@Override
+		public boolean visit(final MethodInvocation node) {
+			if (this.sourceFound == false) {
```
Comment (kruegers) [inline plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/EnsuresPredicateFix.java]: !this.sourceFound

File: plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/EnsuresPredicateFix.java [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+						addMethodEnsuresPredicate(node, this.unit, this.sourceUnit);
+						return false;
+					} catch (JavaModelException | IllegalArgumentException | MalformedTreeException
+							| BadLocationException e) {
+						Activator.getDefault().logError(e);
+					}
+				} else {
+					return true;
+				}
+			} else {
```
Comment (kruegers) [inline plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/EnsuresPredicateFix.java:281]: This else branch is not necessary.

File: plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/EnsuresPredicateFix.java [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+		}
+
+		@Override
+		public boolean visit(final MethodInvocation node) {
+			if (this.sourceFound == false) {
+				if (this.lineNumber == this.unit.getLineNumber(node.getStartPosition())) {
+					try {
+						this.sourceFound = true;
+						addMethodEnsuresPredicate(node, this.unit, this.sourceUnit);
+						return false;
```
Comment (kruegers) [inline plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/EnsuresPredicateFix.java]: This return statement is not necessary.

File: plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/QuickFixer.java [hunk 4]
```diff
@@ -13,17 +15,33 @@
...
 			message = (String) mk.getAttribute(IMarker.MESSAGE);
+			quickFixes.add(new SuppressWarningFix(Constants.SUPPRESSWARNING_FIX + message));
+
+			//TODO: currently the ensuresPredicate quickfix is disabled.
+//			if (errorType.equals(Constants.REQUIRED_PREDICATE_MARKER_TYPE)) {
+//				String predicate = (String) mk.getAttribute("predicate");
+//				
+//				if(predicate.equals("generatedKey") || predicate.equals("randomized")) {
+//					quickFixes.add(new EnsuresPredicateFix("This object comes from a stream/database/other external source and is actually secure."));
+//				}
```
Comment (kruegers) [inline plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/QuickFixer.java]: Is this code needed? If not, please remove.

File: plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/QuickFixer.java [hunk 5]
```diff
@@ -13,17 +15,33 @@
  * @author André Sonntag
  */
 public class QuickFixer implements IMarkerResolutionGenerator {
+	
+	ArrayList<IMarkerResolution> quickFixes;
```
Comment (kruegers) [inline plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/QuickFixer.java]: Please assign all fields a visibility keyword- whatever is the most restrictive possible.

**Review by kruegers — CHANGES_REQUESTED** [review]

**Comment by johspaeth** [comment]
Thanks a lot. 
If the current project is of maven nature, we have to add "CogniCrypt.jar" as a dependency to the pom.xml, instead to the actual eclipse project. Otherwise the project will not compile from command line.

File: plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/QuickFixer.java (hunk 4 above)
Comment (AndreSonntag) [reply plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/QuickFixer.java]: This is the code for enabling the quickfix

Comment (kruegers) [reply plugins/de.cognicrypt.staticanalyzer/src/de/cognicrypt/staticanalyzer/markerresolution/QuickFixer.java]: Why is it commented then?

**Pull request closed without merging**